### PR TITLE
issue43: test_time2_variables 

### PR DESCRIPTION
**Pull request by maggienj**

### CODE REVIEW

File: pyLibrary/queries/es14/aggs.py [hunk 1]
```diff
@@ -159,6 +164,9 @@ def es_aggsop(es, frum, query):
             if s.aggregate == "count":
                 es_query.aggs[literal_field(canonical_name)].value_count.field = field_name
                 s.pull = literal_field(canonical_name) + ".value"
+            if s.aggregate == "sum":
+                es_query.aggs[literal_field(canonical_name)].sum.field = field_name
+                s.pull = literal_field(canonical_name) + ".sum"
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/aggs.py]: Remove this

File: pyLibrary/queries/es14/decoders.py [hunk 2]
```diff
@@ -155,31 +155,57 @@ def append_query(self, es_query, start):
...
-                            field.missing().to_esfilter(),
-                            {"not": {"terms": {field.var: include}}}
-                        ]}},
-                        es_query
-                    ),
-                }})
+                Log.note("decoders.py - before - here is the query")
+                Log.note("{{data}}", data=es_query)
+                if self.limit == 0:
+                    resp = wrap({"aggs": {
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: remove limit==0 blocks

File: pyLibrary/queries/es14/decoders.py [hunk 3]
```diff
@@ -191,21 +217,26 @@ def append_query(self, es_query, start):
                         "include": include
                     }}, es_query),
                     "_missing": set_default(
-                        {"filter": {"or": [
-                            field.missing().to_esfilter(),
+                        {"filter": {"bool": {"should": [
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: use the OrOp(blah blah blah).to_esfilter

File: pyLibrary/queries/es14/decoders.py [hunk 4]
```diff
@@ -191,21 +217,26 @@ def append_query(self, es_query, start):
...
             else:
-                return wrap({"aggs": {
-                    "_match": set_default({"terms": {
-                        "script_field": field.to_ruby(),
-                        "size": self.limit,
-                        "include": include
-                    }}, es_query)
-                }})
+                if self.limit == 0:
+                    return wrap({"aggs": {
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: remove this block

File: pyLibrary/queries/es14/decoders.py [hunk 5]
```diff
@@ -237,13 +268,18 @@ def _range_composer(edge, domain, es_query, to_float):
...
-            ]}},
+            {"filter":
+                 {OrOp("or",
+                        [
+                        OrOp("or",
+                            [
+                            InequalityOp("lt", [edge.value, Literal(None, to_float(_min))]),
+                            InequalityOp("gte", [edge.value, Literal(None, to_float(_max))]),
+                            ]).to_esfilter(),
+                        {"bool": {"must_not": edge.value.exists().to_esfilter()}}
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: Fix this to something like

    OrOp("or", [
        InequalityOp("lt", [edge.value, Literal(None, to_float(_min))]),
        InequalityOp("gte", [edge.value, Literal(None, to_float(_max))]),
        edge.value.missing()
    ] ).to_esfilter()

File: pyLibrary/queries/es14/decoders.py [hunk 6]
```diff
@@ -497,7 +533,7 @@ def append_query(self, es_query, start):
                     }},
                     es_query
                 ),
-                "_missing": set_default({"missing": {"field": self.edge.value}}, es_query)  # TODO: Use Expression.missing().esfilter() TO GET OPTIMIZED FILTER
+                "_missing": set_default( {"bool": {"must_not": self.edge.value.exists().to_esfilter()}}, es_query)  # TODO: Use Expression.missing().esfilter() TO GET OPTIMIZED FILTER
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: change `{"bool": {"must_not": self.edge.value.exists().to_esfilter()}}` to `self.edge.value.missing().to_esfilter()`

File: pyLibrary/queries/es14/decoders.py [hunk 7]
```diff
@@ -164,11 +163,12 @@ def append_query(self, es_query, start):
...
                     }}, es_query),
                     "_missing": set_default(
-                        {"filter": {"or": [
-                            field.missing().to_esfilter(),
-                            {"not": {"terms": {field.var: include}}}
-                        ]}},
-                        es_query
+                        {"filter":
+                             OrOp("or", [
+                                 self.edge.field.missing().to_esfilter(),
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: change from `self.edge.field.missing().to_esfilter(),` to `self.edge.field.missing(),`

File: pyLibrary/queries/es14/decoders.py [hunk 8]
```diff
@@ -235,15 +235,20 @@ def _range_composer(edge, domain, es_query, to_float):
...
         missing_filter = set_default(
-            {"filter": {"or": [
-                OrOp("or", [
-                    InequalityOp("lt", [edge.value, Literal(None, to_float(_min))]),
-                    InequalityOp("gte", [edge.value, Literal(None, to_float(_max))]),
-                ]).to_esfilter(),
-                edge.value.missing().to_esfilter()
-            ]}},
+            {"filter":
+                 {OrOp("or",
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: remove the preceding `{`

File: pyLibrary/queries/es14/decoders.py [hunk 9]
```diff
@@ -472,7 +477,7 @@ def append_query(self, es_query, start):
 
         if not isinstance(self.edge.value, Variable):
             script_field = self.edge.value.to_ruby()
-            missing = self.edge.value.missing()
+            missing = self.edge.value.missing().to_esfilter()
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: remove the `.to_esfilter()` it is called below

File: pyLibrary/queries/es14/decoders.py [hunk 10]
```diff
@@ -574,7 +580,8 @@ def append_query(self, es_query, start):
                 }}, es_query)
             }})
             if self.edge.allowNulls:
-                nest.aggs._missing = set_default({"missing": {"field": v}}, es_query)  # TODO: Use Expression.missing().esfilter() TO GET OPTIMIZED FILTER
+                nest.aggs._missing = set_default(self.edge.field.missing().to_esfilter(),
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: Change to `nest.aggs._missing = set_default(Variable(v).missing().to_esfilter(),`

File: pyLibrary/queries/es14/aggs.py [hunk 11]
```diff
@@ -106,12 +106,17 @@ def get_decoders_by_depth(query):
...
         output[max_depth].append(AggsDecoder(edge, query, limit))
     return output
 
 
 def es_aggsop(es, frum, query):
+    Log.note("aggs.py - params - here are the param values to es_aggsop")
+    Log.note("es {{data}}", data=es)
+    Log.note("frum {{data}}", data=frum)
+    Log.note("query {{data}}", data=query)
+
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/aggs.py]: remove these notes

File: pyLibrary/queries/es14/decoders.py [hunk 12]
```diff
@@ -235,17 +234,16 @@ def _range_composer(edge, domain, es_query, to_float):
...
-                    InequalityOp("lt", [edge.value, Literal(None, to_float(_min))]),
-                    InequalityOp("gte", [edge.value, Literal(None, to_float(_max))]),
-                ]).to_esfilter(),
-                edge.value.missing().to_esfilter()
-            ]}},
-            es_query
-        )
+            {"filter":
+                 OrOp("or", [
+                    OrOp("or", [
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: `{"or":[{"or":[a,b]}, c]}  == {"or":[a,b,c]}`

File: pyLibrary/queries/es14/decoders.py [hunk 13]
```diff
@@ -497,7 +495,8 @@ def append_query(self, es_query, start):
                     }},
                     es_query
                 ),
-                "_missing": set_default({"missing": {"field": self.edge.value}}, es_query)  # TODO: Use Expression.missing().esfilter() TO GET OPTIMIZED FILTER
+                "_missing": set_default(self.edge.value.missing().to_esfilter(), es_query)
+                # TODO: Use Expression.missing().esfilter() TO GET OPTIMIZED FILTER
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: remove this comment, it is done

File: pyLibrary/queries/es14/decoders.py [hunk 14]
```diff
@@ -509,7 +508,8 @@ def append_query(self, es_query, start):
...
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: remove

File: pyLibrary/queries/es14/decoders.py [hunk 15]
```diff
@@ -574,7 +573,8 @@ def append_query(self, es_query, start):
                 }}, es_query)
             }})
             if self.edge.allowNulls:
-                nest.aggs._missing = set_default({"missing": {"field": v}}, es_query)  # TODO: Use Expression.missing().esfilter() TO GET OPTIMIZED FILTER
+                nest.aggs._missing = set_default(Variable(v).missing().to_esfilter(),
+                                                 es_query)  # TODO: Use Expression.missing().esfilter() TO GET OPTIMIZED FILTER
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: remove comment, put all on one one

File: tests/test_jx/test_time_domain.py [hunk 16]
```diff
@@ -165,42 +165,6 @@ def test_time2_variables(self):
...
-                            "interval": DAY.seconds,
-                            "partitions": [{"min": r.t, "max": (Date(r.t) + DAY).unix} for r in expected2 if r.t != None and r.a == "x"]
-                        }
-                    }
-                ],
-                "data": {"v": [
-                    [r.v for r in expected2 if r.a == "x"],
-                    [r.v for r in expected2 if r.a == "y"],
-                    [NULL for r in expected2 if r.a == "x"]
-                ]}
```
Comment (klahnakoski) [inline tests/test_jx/test_time_domain.py]: revert this removal, we need these tests

File: pyLibrary/queries/es14/aggs.py [hunk 17]
```diff
@@ -317,7 +322,11 @@ def es_aggsop(es, frum, query):
             Log.error("Where clause is too deep")
 
     for d in decoders[0]:
+        Log.note("aggs.py - before - here is the es_query")
+        Log.note("{{data}}", data=es_query)
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/aggs.py]: The logging is mutlithreaded; you probably want this on one line so they do not separate.

    Log.note("aggs.py - before - here is the es_query\n{{data}}", data=es_query)

File: pyLibrary/queries/es14/decoders.py [hunk 18]
```diff
@@ -164,11 +162,12 @@ def append_query(self, es_query, start):
                         "order": {"_term": self.sorted} if self.sorted else None
                     }}, es_query),
                     "_missing": set_default(
-                        {"filter": {"or": [
-                            field.missing().to_esfilter(),
-                            {"not": {"terms": {field.var: include}}}
-                        ]}},
-                        es_query
+                        {"filter":
+                             OrOp("or", [
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: If object has just one property ("filter"), then put the value on same line.

    {"filter": OrOp("or", [

File: pyLibrary/queries/es14/aggs.py (hunk 11 above)
Comment (maggienj) [inline pyLibrary/queries/es14/aggs.py]: would like to keep these lines till we complete the code changes for ES-5 compatibility.

File: pyLibrary/queries/es14/decoders.py [hunk 19]
```diff
@@ -235,17 +233,13 @@ def _range_composer(edge, domain, es_query, to_float):
...
-                ]).to_esfilter(),
-                edge.value.missing().to_esfilter()
-            ]}},
-            es_query
-        )
+            {"filter": OrOp("or", [
+                        InequalityOp("lt", [edge.value, Literal(None, to_float(_min))]),
+                        InequalityOp("gte", [edge.value, Literal(None, to_float(_max))]),
+                    edge.value.missing()
+                ]).to_esfilter()}, es_query)
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: ... and reformat

File: pyLibrary/queries/es14/decoders.py [hunk 20]
```diff
@@ -574,7 +567,8 @@ def append_query(self, es_query, start):
                 }}, es_query)
             }})
             if self.edge.allowNulls:
-                nest.aggs._missing = set_default({"missing": {"field": v}}, es_query)  # TODO: Use Expression.missing().esfilter() TO GET OPTIMIZED FILTER
+                nest.aggs._missing = set_default(Variable(v).missing().to_esfilter(),
+                                                 es_query)
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/decoders.py]: put on same line

File: pyLibrary/queries/es14/aggs.py [hunk 21]
```diff
@@ -317,7 +322,10 @@ def es_aggsop(es, frum, query):
             Log.error("Where clause is too deep")
 
     for d in decoders[0]:
+        Log.note("aggs.py - before - here is the es_query\n{{data}}", data=es_query)
         es_query = d.append_query(es_query, start)
+        Log.note("aggs.py - after - here is the es_query")
+        Log.note("{{data}}", data=es_query)
```
Comment (klahnakoski) [inline pyLibrary/queries/es14/aggs.py]: merge to one line, like above

**Pull request merged**